### PR TITLE
Adding multiple layouts to HAL executable creation.

### LINKED ISSUE #4536: Executable linking produces executables with multiple layouts but runtime does not support that
`-iree-hal-link-executables` will create an executable (in LLVM & VMLA) that has multiple interfaces and these then map to multiple executable layouts that are required during binding/dispatch. This doesn't match up with the executable creation semantics which take only a single executable layout - meaning that if the executable layout passed during creation is actually used things get .... weird.

Probably need to change executable creation to take multiple layouts. I've been meaning to change that API for awhile anyway and see if we can make it something a bit more sensible for batch compilation too.

For example, today:
```c
iree_hal_executable_cache_prepare_executable(cache, executable_layout, caching_mode, executable_data, ...);
```

Simple fix:
```c
iree_hal_executable_cache_prepare_executable(cache, caching_mode, executable_data, executable_layout[], ...);
```

Future async coolness:
```c
iree_hal_executable_cache_prepare_executables(cache, caching_mode, executable_data, executable_layout[], signal_semaphore, ...);
```

### PR DESCRIPTION
**Pull request by benvanik**
The fact that we were providing one layout per executable was always a bug and only worked because nothing actually used that argument. This fixes things so that we properly pass a layout per entry point of the executable so that linked executables (which may have any number of interfaces) work correctly.

As a side-effect, this PR also simplifies the HAL interface exposed to the compiler to hide the executable caches. That they were represented at all in the IR was a bit weird as they are a runtime-provider controlled concept and the lifetime of them is critically tied to the lifetime of the loaded compiled module. This removes the hack that was implicitly keeping the executables live in the VM.

It also required minor compiler changes because of the split dispatch regions stuff in the legacy spir-v path and I had to add a workaround to fix up the IR. The existing approach (where the SPIR-V backend didn't tell the HAL about the entry points it created) only worked because we were not validating executable layouts and the vulkan pipelines were just created with the same single layout provided. The upside of this hack is that we could probably start linking SPIR-V modules.

Fixes #4536.

### CODE REVIEW

**Comment by benvanik** [comment]
All tests but lit ones are passing now - will do in the AM. Turns out DCE actually works now because of these fixes and some of the tests are assuming it doesn't :P

File: iree/compiler/Conversion/LinalgToSPIRV/MaterializeEntryPointsPass.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+};  // namespace
+
+std::unique_ptr<OperationPass<ModuleOp>> createMaterializeEntryPointsPass() {
+  return std::make_unique<MaterializeEntryPointsPass>();
+}
+
+static PassRegistration<MaterializeEntryPointsPass> pass(
+    "iree-codegen-materialize-entry-points",
+    "Materialize HAL entry points for each spv.EntryPoint",
```
Comment (ScottTodd) [inline iree/compiler/Conversion/LinalgToSPIRV/MaterializeEntryPointsPass.cpp:72]: Want to add tests for this pass? (fine to skip if the pass will be deleted along with `SplitDispatchFunctionPass` soon)

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/WindowsLinkerTool.cpp [hunk 2]
```diff
@@ -107,6 +107,9 @@ class WindowsLinkerTool : public LinkerTool {
     SmallVector<std::string, 8> flags = {
         getToolPath(),
 
+        // Hide the linker banner message printed each time.
+        "/nologo",
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/internal/WindowsLinkerTool.cpp:111]: 😆

File: iree/compiler/Dialect/HAL/Transforms/MaterializeResourceCaches.cpp [hunk 3]
```diff
@@ -196,101 +179,104 @@ class MaterializeResourceCachesPass
...
-        assert(executableIt != executableCache_.end() &&
-               "executable must have been cached");
-        auto executableVariableOp = executableIt->second;
-
-        // TODO(benvanik): support multiple interfaces. We'd probably want to
-        // store each executable+interface as a variable.
-        //
-        // This is *only* safe now because any backends that support multiple
-        // interfaces during compilation do *not* use layouts during executable
-        // cache preparation.
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Transforms/MaterializeResourceCaches.cpp:273]: So happy to see this resolved :D

**Review by ScottTodd — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/MaterializeEntryPointsPass.cpp (hunk 1 above)
Comment (benvanik) [reply iree/compiler/Conversion/LinalgToSPIRV/MaterializeEntryPointsPass.cpp:72]: yeah this entire path is woefully undertested - there were only two or three tests in all of IREE that failed with these multiple spir-v entry points (simple_embedding_test was one of them lol). Really hoping we can kill it off soon, and I think the work mahesh is doing is going to make that very possible!

**Pull request merged**